apollo-router/src/configuration/cors.rs

```diff
@@ -84,90 +154,338 @@ impl Cors {
+        }
+
+        // Validate global methods
+        parse_values::<http::Method>(&config.methods, "method")?;
+
+        // Validate global expose headers
+        if let Some(headers) = &config.expose_headers {
+            parse_values::<http::HeaderName>(headers, "expose header name")?;
+        }
+
+        // Validate origin configurations
+        for origin_config in &config.origins {
+            // Validate origin URLs
+            for origin in &origin_config.origins {
+                http::HeaderValue::from_str(origin).map_err(|_| {
+                    format!(
+                        "origin '{}' is not valid: failed to parse header value",
+                        origin
+                    )
+                })?;
+            }
+
+            // Validate origin-specific headers
+            if !origin_config.allow_headers.is_empty() {
+                parse_values::<http::HeaderName>(
+                    &origin_config.allow_headers,
+                    "allow header name",
+                )?;
+            }
+
+            // Validate origin-specific methods
+            if !origin_config.methods.is_empty() {
+                parse_values::<http::Method>(&origin_config.methods, "method")?;
+            }
+
+            // Validate origin-specific expose headers
+            if !origin_config.expose_headers.is_empty() {
+                parse_values::<http::HeaderName>(
+                    &origin_config.expose_headers,
+                    "expose header name",
+                )?;
+            }
+        }
+
+        Ok(Self { config })
+    }
+}
+
+impl<S> Layer<S> for CorsLayer {
+    type Service = CorsService<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        CorsService {
+            inner: service,
+            config: self.config.clone(),
+        }
+    }
+}
+
+/// Our custom CORS service that handles per-origin configuration
+#[derive(Clone, Debug)]
+pub(crate) struct CorsService<S> {
+    inner: S,
+    config: Cors,
+}
+
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for CorsService<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>> + Send + 'static,
+    S::Future: Send + 'static,
+    S::Error: Send + 'static,
+    ReqBody: Send + 'static,
+    ResBody: Send + 'static + Default,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let origin = req.headers().get(ORIGIN).cloned();
+        let is_preflight = req.method() == http::Method::OPTIONS;
+        let config = self.config.clone();
+        let request_headers = req.headers().get("access-control-request-headers").cloned();
```

Comment:
There is a constant for this:
```suggestion
        let request_headers = req.headers().get(ACCESS_CONTROL_REQUEST_HEADERS).cloned();
```